Add OASys import risk information screens to Apply from AP (Part 2)

bryangaledxw opened the pull request:
# Changes in this PR

Following from #203, this PR adds the Apply pages responsible for actually fetching data from OASys, with the supporting information page using our input to the optional OASys sections page

## Screenshots of UI changes

### Before
![223161148-2b9bb0fa-498e-4d42-972a-beb70076aba9](https://user-images.githubusercontent.com/94137563/223181631-99ce22c7-14b0-4d3e-aa6a-afc0fc59ad9d.png)


### After
![localhost_3000_applications_b826b0ca-8e28-4d80-ba14-f3765b68ced0_tasks_oasys-import_pages_rosh-summary](https://user-images.githubusercontent.com/94137563/223180939-69828126-e84f-4071-aded-c42c37582f4a.png)
![localhost_3000_applications_b826b0ca-8e28-4d80-ba14-f3765b68ced0_tasks_check-your-answers_pages_review](https://user-images.githubusercontent.com/94137563/223180958-7f55e8b5-5cb6-416c-bd15-22f9711249df.png)

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

bryangaledxw committed:
Add RoshSummary Apply form page

We directly copy the RoshSummary page from Approved Premises

bryangaledxw committed:
Add OffenceDetails Apply form page

We directly copy the OffenceDetails page from Approved Premises

bryangaledxw committed:
Add SupportingInfo Apply form page

We directly copy the SupportingInfo page from Approved Premises

bryangaledxw committed:
Add RiskManagementPlan Apply form page

We directly copy the RiskManagementPlan page from Approved Premises

bryangaledxw committed:
Add RiskToSelf Apply form page

We directly copy the RiskToSelf page from Approved Premises

bryangaledxw committed:
Add OASys import shared template

We add a Nunjucks template that will be shared between each of our OASys import pages

bryangaledxw committed:
Integrate OASys import pages into Apply journey

We set our new RoshSummary page to be the next page following the optional OASys sections page, integrating our new pages into our Apply journey. We now update our integration tests to complete our new form pages

bryangaledxw committed:
Remove 'convert' prefix from dateUtils functions

In dateUtils, we remove the 'convert' prefix from function names to follow Approved Premises naming. This will make make it more straightforward when using code from Approved Premises that uses dateUtils

bryangaledxw committed:
Sort imports

bryangaledxw committed:
Add completeOasysImportQuestions test helper

We add this helper function to our ApplyPage class, which we will use for testing each of the OASys import pages

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add OAsys import/risk information screens to Apply from AP](https://trello.com/c/1wg76KUW/861-add-oasys-import-risk-information-screens-to-apply-from-ap)

bryangaledxw committed:
Expand OASys related endpoints stubs

We expand the stubbing of OASys related endpoints, that we will need to test our OASys import pages

patrickjfl commented on cypress_shared/pages/apply/applyPage.ts:
```diff
@@ -52,4 +52,13 @@ export default class ApplyPage extends Page {
     cy.get('.govuk-back-link').should('have.attr', 'href').and('include', path)
   }
+
+  completeOasysImportQuestions(section, sectionName: string): void {
```
I'm a bit confused about what a section is here, partly I think because of the absence of a type. If we can't use a type could we make it a bit clearer what it is with our naming? E.g. it needs to be an array, but is singular...?

bryangaledxw replied:
Yep, changed to:
```
  completeOasysImportQuestions(questions: Array<OASysQuestion>, sectionName: string): void {
    questions.forEach(question => {
```

patrickjfl commented on cypress_shared/pages/apply/applyPage.ts:
```diff
@@ -52,4 +52,13 @@ export default class ApplyPage extends Page {
   checkForBackButton(path: string) {
```
Repetition of 'will' in the commit message here ('Add completeOasysImportQuestions test helper')

bryangaledxw replied:
✅

patrickjfl commented on server/form-pages/apply/risk-and-need-factors/oasys-import/roshSummary.test.ts:
```diff
@@ -0,0 +1,85 @@
+  const personRisks = risksFactory.build()
+  const application = applicationFactory.build({ risks: personRisks })
+
+  describe('initialize', () => {
```
annoying that we aren't standardised on 'initialise' spelling across the project!

bryangaledxw replied:
You have my blessing to do a global find and replace!

patrickjfl reviewed (approved):
Looks good! Just that test helper function that I think could be a bit clearer.

bryangaledxw committed:
Use feature flag to disable Apply feature

Following the lead of our new voids feature, we disable Apply via configuration in config.ts, rather than checking the environment in our routes setup

bryangaledxw commented:
@patrickjfl Mind having a quick look at the second commit? Just bringing things more in line with voids

patrickjfl reviewed (approved):
LGTM 👍
